libvast/include/vast/table_slice.hpp

```diff
@@ -314,23 +283,49 @@ class table_slice final {
 
 // -- operations ---------------------------------------------------------------
 
+/// Concatenates all slices in the given range.
+/// @param slices The input table slices.
+table_slice concatenate(std::vector<table_slice> slices);
+
 /// Selects all rows in `slice` with event IDs in `selection`. Cuts `slice`
 /// into multiple slices if `selection` produces gaps.
 /// @param slice The input table slice.
-/// @param selection ID set for selecting events from `slice`.
-/// @returns new table slices of the same implementation type as `slice` from
-///          `selection`.
+/// @param expr The filter expression.
+/// @param hints ID set for selecting events from `slice`.
 /// @pre `slice.encoding() != table_slice_encoding::none`
-std::vector<table_slice> select(const table_slice& slice, const ids& selection);
+detail::generator<table_slice>
+select(const table_slice& slice, expression expr, const ids& hints);
+
+/// Produces a new table slice consisting only of events addressed in `hints`
+/// that match the given expression. Does not preserve ids; use `select`
+/// instead if the id mapping must be maintained.
+/// @param slice The input table slice.
+/// @param expr The expression to evaluate.
+/// @param hints An ID set for pruning the events that need to be considered.
+/// @returns a new table slice consisting only of events matching the given
+///          expression.
+/// @pre `slice.encoding() != table_slice_encoding::none`
+std::optional<table_slice>
+filter(const table_slice& slice, expression expr, const ids& hints);
```

Comment:
Meta-comment: If this is now much faster than before than we should write a changelog entry.

Reply:
I would argue that performance-related changes should rather go in a release blog post.

Reply:
Makes sense to me.